context/llpcCompiler.cpp

```diff
@@ -823,6 +832,170 @@ Result Compiler::BuildShaderModule(
     return result;
 }
 
+Result Compiler::BuildPipelineUsingRelocatableElf(
+    Context*                            pContext,                   // [in] Acquired context
+    ArrayRef<const PipelineShaderInfo*> shaderInfo,                 // [in] Shader info of this pipeline
+    uint32_t                            forceLoopUnrollCount,       // [in] Force loop unroll count (0 means disable)
+    ElfPackage*                         pPipelineElf)               // [out] Output Elf package
+{
+    Result result = Result::Success;
+
+    // Determine whether or not relocatable elf can be used.
+    // We only do relocatable elf if it is a VsPs shader that is given in
+    // spir-v.
+    bool useRelocatableShaders = true;
+    for (uint32_t stage = 0; stage < shaderInfo.size(); ++stage)
+    {
+        if (stage != ShaderStageVertex && stage != ShaderStageFragment)
+        {
+            if ((shaderInfo[stage] != nullptr) && (shaderInfo[stage]->pModuleData != nullptr))
+            {
+                useRelocatableShaders = false;
+            }
+        }
+        else if (shaderInfo[stage] == nullptr || shaderInfo[stage]->pModuleData == nullptr)
+        {
+            useRelocatableShaders = false;
+        }
+    }
+
+    if (useRelocatableShaders && shaderInfo[0] != nullptr)
+    {
+        const ShaderModuleData* pModuleData = reinterpret_cast<const ShaderModuleData*>(shaderInfo[0]->pModuleData);
+        if ((pModuleData != nullptr) && (pModuleData->binType == BinaryType::LlvmBc))
+        {
+          useRelocatableShaders = false;
+        }
+    }
+
+    // For debugging you can add the option |-relocatable-shader-elf-limit=<n>|
+    // to limit the number of time relocatable shaders are used to n.
+    if (useRelocatableShaders && cl::RelocatableLimit != -1) {
+      static uint32_t counter = 0;
+      if (counter >= cl::RelocatableLimit) {
+        useRelocatableShaders = false;
+      } else {
+        ++counter;
+      }
+    }
+
+    if (!useRelocatableShaders) {
```

Comment:
So... there's a !useRelocatableShaders path in a function called BuildPipelineUsingRelocatableElf.

That seems wrong. Plus, I suspect there's a lot of code duplication here with other things.

Reply:
I agree we should avoid duplicated code. can we use following logic
if (CheckCanUseRelocatableShaderElf())
       BuildPipelineUsingRelocatableElf()
else
      BuildPipelineInternal()

Reply:
Very good idea. I'll do that.